src/backend/egl/fence.rs

```diff
@@ -0,0 +1,163 @@
+//! EGL fence related structs
+
+use std::os::unix::io::{AsRawFd, FromRawFd, IntoRawFd, OwnedFd};
+use std::sync::Arc;
+use std::time::Duration;
+
+use crate::backend::egl::{ffi, wrap_egl_call, EGLDisplay, EGLDisplayHandle, Error};
+
+#[derive(Debug)]
+struct InnerEGLFence {
+    display_handle: Arc<EGLDisplayHandle>,
+    handle: ffi::egl::types::EGLSync,
+    native: bool,
+}
+unsafe impl Send for InnerEGLFence {}
+unsafe impl Sync for InnerEGLFence {}
+
+impl Drop for InnerEGLFence {
+    fn drop(&mut self) {
+        unsafe {
+            // Not much we can do about when this fails
+            ffi::egl::DestroySync(**self.display_handle, self.handle);
+        }
+    }
+}
+
+/// A EGL fence
+#[derive(Debug, Clone)]
+pub struct EGLFence(Arc<InnerEGLFence>);
+
+impl EGLFence {
+    /// Returns whether the display supports creating fences from native fences
+    pub fn supports_from_native(display: &EGLDisplay) -> bool {
+        display.supports_native_fences
+    }
+
+    /// Create a fence from a native fence
+    pub fn from_native(display: &EGLDisplay, native: OwnedFd) -> Result<Self, Error> {
+        // SAFETY: we do not have to test for EGL_KHR_fence_sync as EGL_ANDROID_native_fence_sync
+        // requires it already
+        if !display.supports_native_fences {
+            return Err(Error::EglExtensionNotSupported(&[
+                "EGL_ANDROID_native_fence_sync",
+            ]));
+        }
+
+        let display_handle = display.get_display_handle();
+        let attributes: [ffi::egl::types::EGLAttrib; 3] = [
+            ffi::egl::SYNC_NATIVE_FENCE_FD_ANDROID as ffi::egl::types::EGLAttrib,
+            native.as_raw_fd() as ffi::egl::types::EGLAttrib,
+            ffi::egl::NONE as ffi::egl::types::EGLAttrib,
+        ];
+        let handle = wrap_egl_call(|| unsafe {
+            ffi::egl::CreateSync(
+                **display_handle,
+                ffi::egl::SYNC_NATIVE_FENCE_ANDROID,
+                attributes.as_ptr(),
+            )
+        })
+        .map_err(Error::CreationFailed)?;
+        // SAFETY: Ownership has been transferred to egl
```

Comment:
Because EGL took ownership of the fd, it would probably be more clear to use mem::forget

Reply:
In case of an error we want to make sure we close the fd, but yes, I can restructure that a bit to use `into_raw_fd` and close the fd in case of an error.